File: April-18/11-04-18/python/02-Prime-Numbers-Again.py

prime = []
N = 10000
dp = [0]*(N+1)
def seive():
    check = [0]*(N+1);
    for i in range(2, N+1):
        if(check[i] == 0):
            prime.append(i)
            for j in range(2*i,N+1,i):
                check[j] = 1
            dp[i] = 1
            if(i <= 5):
                val = pow(i,i)
                dp[val] = 1
def solve():
    dp[0] = dp[1] = N
    for i in range(2,N+1):
        if(dp[i] != 1):
            dp[i] = N
            for val in prime:
                if(val > i):
                    break
                if(i-val >= 0 ):
                    dp[i] = min(dp[i-val] + 1, dp[i])
                if(val <= 5):
                    val = pow(val,val)
                    if(i-val >= 0):
                        dp[i] = min(dp[i-val] + 1, dp[i])
    
seive()
solve()
for _ in range(int(input())):
    n = int(input())
    print(dp[n])

File: April-18/11-04-18/java/02-Prime-Numbers-Again.java


import java.util.*;

class TestClass {
    static Vector vec = new Vector();
    static int dp[] = new int[10001];
    static void seive()
    {
        int n = 10000;
        int check[] = new int[n+1];
        for(int i = 2; i <= n; i++)
        {
            if(check[i] == 0)
            {
                vec.add(i);
                for(int j = 2*i; j <= n; j += i)
                {
                    check[j] = 1;
                }
                dp[i] = 1;
                if(i <= 5)
                {
                    int val = (int)Math.pow(i,i);
                    dp[val] = 1;
                }
            }
        }
    }
    
    static void solve()
    {
        int n = 10000;
        dp[0] = dp[1] = n+1;
        for(int i = 2; i <= n; i++)
        {
            if(dp[i] != 1)
            {
                dp[i] = n+1;
                Enumeration vEnum = vec.elements();
                while(vEnum.hasMoreElements())
                {   
                    int val = (int)vEnum.nextElement();
                    if(val > i)
                        break;
                    if(i-val >= 0)
                        dp[i] = Math.min(dp[i-val] + 1, dp[i]);
                    if(val <= 5)
                    {
                        val = (int)Math.pow(val,val);
                        if(i-val >=0 )
                            dp[i] = Math.min(dp[i-val] + 1, dp[i]);
                    }
                    
                }
            }
        }
    }
    public static void main(String args[] ) throws Exception {
        seive();
        solve();
        Scanner input = new Scanner(System.in);
        int t = input.nextInt();
        while(t-- > 0)
        {
            int n = input.nextInt();
            System.out.println(dp[n]);
        }
    }
}
